### PR TITLE
LPS-151251 Create URLBuilder to prep converting HttpComponentsUtil param methods into builder pattern

### PR DESCRIPTION
**Pull request by julschong**
@tinatian Hi Tina,

this is resubmitted #1041 

Main Task: [LPS-150839 Remove petra-apache-http-components](https://issues.liferay.com/browse/LPS-150839)
Sub Task: [LPS-151251 Create URLBuilder interface to provide builder interface to HttpComponentUtil](https://issues.liferay.com/browse/LPS-151251)

URLBuilder has no logic currently. Only using HttpComponentsUtil's logic.
Next part for this task is to move the logic to URLBuilder and apply to usages.

Additional note for this pull:

1. URLBuilder now save each operation and only modify _url inside build() method. This change is to follow conventional Builder pattern for easier understanding.
2. operations are cleared after build()
3. Added a test case to make sure operational sequence is followed

Thank you!

### CODE REVIEW

**Comment by julschong** [comment]
@catarineaguiar Hi Catarine, could you please help me scan this test result?
Thank you!

**Comment by catarineaguiar** [comment]
@julschong no regressions

**Comment by julschong** [comment]
@tinatian 
Hi Tina,

Could you please help me review this pull please?

Thank you!

File: portal-kernel/src/com/liferay/portal/kernel/url/URLBuilder.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	private URLBuilder(String url) {
+		_url = url;
+	}
+
+	private String _url;
+	private final List<URLOperation> _urlOperations = new LinkedList<>();
+
+	private static class URLOperation {
+
+		public URLOperation(
```
Comment (tinatian) [inline portal-kernel/src/com/liferay/portal/kernel/url/URLBuilder.java:134]: @julschong Please use private constructor.

**Comment by tinatian** [comment]
@julschong Please apply the old usages of old URLBuilder, and remove the old one in the same pull.

TIna

**Pull request closed without merging**